### PR TITLE
Feat/5853 user entity mandatory

### PR DESCRIPTION
**Pull request by morhag90**
Close: #5853 

## PR Details

Add User entity validation before plugin installation. 

### Summary
🔒🛠️🔄

The pull request adds validation and error handling to prevent installing authentication plugins without a user entity in the Amplication platform. It also fixes some bugs with plugin settings and configurations, and enables circular dependency injection between the plugin installation and the resource modules and services. The changes affect the `PluginsCatalog` component, the `pluginInstallation` and `resource` modules and services, and the `GET_RESOURCES` query.

> _Oh we're the crew of the Amplication ship_
> _And we code with skill and care_
> _We add some fields and validations_
> _To the `GET_RESOURCES` query there_

### Walkthrough
*  Add a dialog component to the PluginsCatalog component to show a message when the user tries to install an authentication plugin without having a user entity ([link](https://github.com/amplication/amplication/pull/6123/files?diff=unified&w=0#diff-c04b321b9fbe9466d4437357cde44fb0d2d37ff0f142837eadadc332969f99cbL1-R10), [link](https://github.com/amplication/amplication/pull/6123/files?diff=unified&w=0#diff-c04b321b9fbe9466d4437357cde44fb0d2d37ff0f142837eadadc332969f99cbL16-R31), [link](https://github.com/amplication/amplication/pull/6123/files?diff=unified&w=0#diff-c04b321b9fbe9466d4437357cde44fb0d2d37ff0f142837eadadc332969f99cbL49-R76), [link](https://github.com/amplication/amplication/pull/6123/files?diff=unified&w=0#diff-c04b321b9fbe9466d4437357cde44fb0d2d37ff0f142837eadadc332969f99cbR107-R126))
*  Modify the handleInstall function of the PluginsCatalog component to check the plugin version configuration and the current resource entities before creating the plugin installation ([link](https://github.com/amplication/amplication/pull/6123/files?diff=unified&w=0#diff-c04b321b9fbe9466d4437357cde44fb0d2d37ff0f142837eadadc332969f99cbL33-R54), [link](https://github.com/amplication/amplication/pull/6123/files?diff=unified&w=0#diff-c04b321b9fbe9466d4437357cde44fb0d2d37ff0f142837eadadc332969f99cbL43-R64))
*  Add a query field to get the entities of each resource in the GET_RESOURCES query ([link](https://github.com/amplication/amplication/pull/6123/files?diff=unified&w=0#diff-fe78680b0515c3532110dca76f1ecea3d62ebd6107d7e04e6f3539b502be06d9R29-R32))
*  Import the resource module and service in the plugin installation module and service using forwardRef and Inject functions to enable circular dependency injection ([link](https://github.com/amplication/amplication/pull/6123/files?diff=unified&w=0#diff-2d4ca8bf657e19322b4e68fbefe72ef42eacd29db38cabbb695ead1712551255L1-R1), [link](https://github.com/amplication/amplication/pull/6123/files?diff=unified&w=0#diff-2d4ca8bf657e19322b4e68fbefe72ef42eacd29db38cabbb695ead1712551255L7-R10), [link](https://github.com/amplication/amplication/pull/6123/files?diff=unified&w=0#diff-8a3c10918f6040ee543eff002f476c2c309524582e12532aae5c8a628cf64ec6L1-R1), [link](https://github.com/amplication/amplication/pull/6123/files?diff=unified&w=0#diff-8a3c10918f6040ee543eff002f476c2c309524582e12532aae5c8a628cf64ec6R16), [link](https://github.com/amplication/amplication/pull/6123/files?diff=unified&w=0#diff-8a3c10918f6040ee543eff002f476c2c309524582e12532aae5c8a628cf64ec6R63-R64))
*  Add a user entity validation method to the resource service that throws a conflict exception if the resource does not have a user entity and the plugin configurations require an authentication entity ([link](https://github.com/amplication/amplication/pull/6123/files?diff=unified&w=0#diff-69e1f5c184f8ffa46def579dd2fb6e88ea5b228afc3ea64079b51134af116c80L7-R12), [link](https://github.com/amplication/amplication/pull/6123/files?diff=unified&w=0#diff-69e1f5c184f8ffa46def579dd2fb6e88ea5b228afc3ea64079b51134af116c80R60), [link](https://github.com/amplication/amplication/pull/6123/files?diff=unified&w=0#diff-69e1f5c184f8ffa46def579dd2fb6e88ea5b228afc3ea64079b51134af116c80R309-R333))
*  Call the user entity validation method of the resource service in the plugin installation service before creating the plugin installation ([link](https://github.com/amplication/amplication/pull/6123/files?diff=unified&w=0#diff-8a3c10918f6040ee543eff002f476c2c309524582e12532aae5c8a628cf64ec6R105-R111), [link](https://github.com/amplication/amplication/pull/6123/files?diff=unified&w=0#diff-69e1f5c184f8ffa46def579dd2fb6e88ea5b228afc3ea64079b51134af116c80L416-R455))

## PR Checklist

- [x] Tests for the changes have been added
- [x] `npm test` doesn't throw any error

### CODE REVIEW

File: packages/amplication-client/src/Plugins/PluginsCatalog.tsx [hunk 1]
```diff
@@ -13,9 +19,15 @@ type Props = AppRouteProps & {
...
 };
-
+const DIALOG_CLASS_NAME = "limitation-dialog";
 const PluginsCatalog: React.FC<Props> = ({ match }: Props) => {
   const { resource } = match.params;
+  const { currentResource } = useContext(AppContext);
+  const [confirmInstall, setConfirmInstall] = useState<boolean>(false);
+
+  const userEntity = currentResource.entities.find(
+    (entity) => entity.name.toLowerCase() === "user"
```
Comment (mulygottlieb) [inline packages/amplication-client/src/Plugins/PluginsCatalog.tsx:29]: Not sure we need to change this now, but remember that a coming task is to allow renaming the user entity so this hard-coded "user" can be problematic later. We can change it when we handle that rename task.

File: packages/amplication-client/src/Plugins/PluginsCatalog.tsx [hunk 2]
```diff
@@ -30,7 +42,15 @@ const PluginsCatalog: React.FC<Props> = ({ match }: Props) => {
   const handleInstall = useCallback(
     (plugin: Plugin, pluginVersion: PluginVersion) => {
       const { name, pluginId, npm } = plugin;
-      const { version, settings } = pluginVersion;
+      const { version, settings, configurations } = pluginVersion;
+      const requireAuthenticationEntity = configurations
+        ? configurations["requireAuthenticationEntity"]
```
Comment (mulygottlieb) [inline packages/amplication-client/src/Plugins/PluginsCatalog.tsx:47]: What if there is a configurations object (not null) but the configuration does not contain a "requireAuthenticationEntity" key? 
That should be optional too.

File: packages/amplication-client/src/Plugins/PluginsCatalog.tsx (hunk 1 above)
Comment (morhag90) [reply packages/amplication-client/src/Plugins/PluginsCatalog.tsx:29]: just for now, to find  the user entity, still need to decide what will be the logic/ configuration

Comment (mulygottlieb) [reply packages/amplication-client/src/Plugins/PluginsCatalog.tsx:29]: Okay, let's just move the string "user" to a const like in the server where you have USER_ENTITY_NAME

File: packages/amplication-client/src/Plugins/PluginsCatalog.tsx (hunk 2 above)
Comment (mulygottlieb) [inline packages/amplication-client/src/Plugins/PluginsCatalog.tsx:47]: Let's change all places the use hard coded "requireAuthenticationEntity" to a const string e.g. const REQUIRE_AUTH_ENTITY = "requireAuthenticationEntity"

**Review by mulygottlieb — APPROVED** [review]
Approving. Just please add a task/reminder to check what happens when a plugin has a configuration, but this configuration does not have the "requireAuthenticationEntity" key (in PluginsCatalog.tsx line 50)

**Pull request merged**